server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -322,7 +325,13 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
   {
     try {
       log.info("Loading segment %s", segment.getIdentifier());
-      if (loadSegment(segment, callback)) {
+      if (segmentsToDelete.contains(segment)) {
+        synchronized (lock) {
+          segmentsToDelete.remove(segment);
+        }
+      }
```

Comment:
@nishantmonu51 Yeah, contains and remove can definitely be moved inside the synchronized block. However, in that case, each time when ```addSegment()``` is called, it has to wait for the lock in order to make progress, which will make things slow. Given that in most cases ```segmentsToDelete.contains(segment)``` returns false, it will save a lot of cost of acquiring lock by doing the "contains" check outside the synchronized block.